File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cs
using System;
using System.Collections.Generic;

public class Solution {
    private void Backtrack(string combination, string digits, int index, string[] phone, IList<string> result) {
        if (index == digits.Length) {
            result.Add(combination);
            return;
        }

        string letters = phone[digits[index] - '0'];
        for (int i = 0; i < letters.Length; i++) {
            Backtrack(combination + letters[i], digits, index + 1, phone, result);
        }
    }

    public IList<string> LetterCombinations(string digits) {
        if (string.IsNullOrEmpty(digits)) {
            return new List<string>();
        }

        string[] phone = {
            "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        List<string> result = new List<string>();
        Backtrack("", digits, 0, phone, result);
        return result;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
function letterCombinations(digits) {
    if (digits.length === 0) {
        return [];
    }

    const phone = [
        '0', '1', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz'
    ];

    const result = [];

    function backtrack(combination, index) {
        if (index === digits.length) {
            result.push(combination);
            return;
        }

        const letters = phone[digits[index]];
        for (let i = 0; i < letters.length; i++) {
            backtrack(combination + letters[i], index + 1);
        }
    }

    backtrack('', 0);
    return result;
}
